Attachments with DQL

ErikEverson opened the pull request:
Build off of https://github.com/getditto/template-app-nodejs-todo/pull/1 and added DQL attachments.
I also added a .gitignore so that files are not automatically pulled in on a git add

This required Ditto 4.6+ to function

Commit:
Added attachments with DQL querys

Commit:
added gitignore to project

Commit:
Updated to not have special attachment case replaced with comment

pvditto commented on index.ts:
```diff
@@ -1,82 +1,173 @@
+    "SELECT * FROM COLLECTION attachments (my_attachment ATTACHMENT) WHERE isDeleted = false AND isCompleted = false",
+    (result) => {
+      attachments = result.items;
+      // Attachments dont seem to show up as part of a document
+      //   .map((doc) => {
+      //   return doc.value;
+      // });
```
This query will return attachment tokens, which can then be used to fetch attachments with `const attachment = await ditto.store.fetchAttachment(result.items[i].value.my_attachment)`. 

Does this query really not return values?

pvditto replied:
Ah, sorry, I see it's there in another part of the script, was just reacting to the commented out bit.

kitplummer replied:
Assuming we do want to update the template app, @ErikEverson - can you remove the commented out block?

ErikEverson replied:
Yeah I will remove that. We will probably have to wait until 4.6 is released first as the code will not compile until 4.6 can be used.

ErikEverson replied:
Code has been cleaned up. Should be ready for merged once 4.6 is done.

Commit:
Cleaned up DQL and removed comment

ErikEverson commented:
Waiting on release of 4.6
- 4.6 has been released

Commit:
updated to 4.6

pvditto commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+ditto/
+files/
+index.js
```
Personally, I found it unexpected that `npm start` runs the build output and doesn't build on the fly and I would change that, e.g. use a `build/` dir for the building, put that on .gitignore, and then change `npm start` to use `ts-node` or similar.

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
-let subscription
-let liveQuery
-let tasks: Document[] = []
+// Attachments dont to up as part of a Document.
+// Using QueryResultItem item instead.
```
Leftover comment

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
+    appID: "REPLACE_WITH_YOUR_APP_ID",
+    token: "REPLACE_WITH_YOUR_TOKEN",
+  });
+  ditto.startSync();
```
```suggestion
  ditto.disableSyncWithV3();
  ditto.startSync();
```

Should always be called when mutating DQL APIs are used.

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
+      console.log("Result from execute:");
+      console.log(result);
+
+      const token = result.items[0].value.my_attachment;
```
result.items may be empty

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
+      console.log("First attachments data:");
+      console.log(attachmentData);
+
+      // Get attchment from the observer as well.
```
```suggestion
      // Get attachment from the observer as well.
```

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
+
+    if (answer.startsWith("--attachments")) {
+      // Fetch the attachment token from a document in the store
+      // is you want to search by ID: "WHERE _id = '123'" syntax is very important single quote only.
```
```suggestion
```

Commit:
added event handler for one of the attachment requests

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
+      const metadata = { name: "image.png" };
+
+      // Copy the file into Ditto's store and create an attachment object.
+      const myAttachment = await ditto.store.newAttachment(path, metadata);
```
needs some error handling probably, at least for code `store/attachment-file-not-found`

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
+
+      // Copy the file into Ditto's store and create an attachment object.
+      const myAttachment = await ditto.store.newAttachment(path, metadata);
+      console.log(myAttachment);
```
This doesn't pretty print unfortunately, is that intentional?

ErikEverson replied:
No it was for debugging while I worked on getting attachments functional. I will remove and clean up

pvditto commented on index.ts:
```diff
@@ -1,82 +1,209 @@
+    }
+
+    if (answer.startsWith("--exit")) {
+      ditto.stopSync();
```
```suggestion
      await ditto.close()
```

pvditto reviewed (commented):
Gave the script a try and adding some things I noticed here.

Commit:
updated .gitignore to not ignore files dir

Commit:
added files and filesOut dirs for demo

Commit:
updated attachment code to copy files over.

Commit:
consolidated command naming

Commit:
Added help and error handling

Commit:
Added more error handling.

Commit:
Added comment for help menu

Commit:
disabled sync with v3

Commit:
Changed to use build dir

Commit:
removed some log calls in async context.

Commit:
DIsabled sync

Commit:
fixed type

kitplummer reviewed (approved):
LGTM, but think @pvditto - should be final reviewer.

pvditto commented on package.json:
```diff
@@ -5,12 +5,12 @@
   "scripts": {
     "build": "tsc",
-    "start": "node index.js"
+    "start": "node build/index.js"
```
To not require users to run both `build` and `start` we could 

```suggestion
    "start": "npm build && node build/index.js"
```

Although that doesn't work on Windows, which lacks `&&` support. Another alternative would be adding `ts-node` to the package so building is not required anymore:

```suggestion
    "start": "ts-node index.ts"
```

Can also stay as it is, I just wanted to make some suggestions.

ErikEverson replied:
I like that I will go with the ts-node as more compatibility is never a bad thing.

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
-import { stdin as input, stdout as output } from 'node:process';
+let ditto: Ditto;
+let taskQueryResult;
+let attachmentQueryResult;
```
```suggestion
let taskQueryResult: QueryResult;
let attachmentQueryResult: QueryResult;
```

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+  console.log("'--insert-task myNewTask'");
+  console.log("'--insert-attachment myFileName'");
+  console.log("   Inserts a task/attachment");
+  console.log("   Inserts an attachment from ./files with the provided name.");
```
There doesn't seem to be any linking between tasks and attachments. Is that intentional? I would expect attachments to be associated with tasks.

ErikEverson replied:
It was intentional for now. I will think about something to try to connect them

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+  console.log("   Inserts a task/attachment");
+  console.log("   Inserts an attachment from ./files with the provided name.");
+  console.log('   Example: "--insert-task Get Milk"');
+  console.log('   Example: "--insert-attachment i.png" adds "./files/i.png"');
```
```suggestion
  console.log('   Example: "--insert-attachment peers.webp" adds "./files/peers.webp"');
```

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+    token: "REPLACE_WITH_YOUR_TOKEN",
+  });
+  
+  ditto.disableSyncWithV3();
```
```suggestion
  await ditto.disableSyncWithV3();
```

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+    // --------------------- Attachments section ---------------------
+
+    if (answer.startsWith("--attachments")) {
+      // Get attachment from the observer and copy all files to the filesOut directory.
```
It could be helpful to document why this is using the observer instead of fetching the list of attachments using `execute`.

ErikEverson replied:
We could use execute. I guess what is the preferred method?
I used to have both but it was a little redundant.

okdistribute replied:
We typically advise developers not to use execute because there's possibility for race conditions

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+            case "Completed":
+              const fetchedAttachment = attachmentFetchEvent.attachment;
+              const name = fetchedAttachment.metadata["name"];
+              fetchedAttachment.copyToPath("./filesOut/" + name);
```
```suggestion
              fetchedAttachment.copyToPath("./filesOut/" + name);
              console.log(`Attachment fetch completed at ./filesOut/${name}`);
```

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+              const name = fetchedAttachment.metadata["name"];
+              fetchedAttachment.copyToPath("./filesOut/" + name);
+              break;
+            default:
```
This `default` case should be changed. `attachmentFetchEvent.type` could also contain information about fetch progress for large attachments.

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+    // --------------------- Exit section ---------------------
+
+    if (answer.startsWith("--exit")) {
+      await ditto.close;
```
```suggestion
      await ditto.close();
```

pvditto commented on index.ts:
```diff
@@ -1,82 +1,289 @@
+        // Handle Errors
+        console.log("An error occured!!!");
+        console.log(e);
       }
```
```suggestion
      } catch(e) {
        if (e instanceof DittoError && e.code === 'store/attachment-file-not-found') {
            console.error(`File not found: ${name}`)
        } else {
            console.error('Error trying to insert attachment:');
            console.error(e);
        }
      }
```

pvditto reviewed (changes requested):
Sorry for taking a while to review this, I was caught up in sprinting to get the on-connecting APIs implemented.

I made a couple of suggestions, the most important would be:
- the attachment fetcher callback has a couple of event types, this should be considered because currently everything other than "download completed" is recognized as an error.
- `disableSyncWithV3` needs to be awaited, I think this comes from my own previous review
- there's a syntax error with `ditto.close()`

In general I think it would be nice if there was some meaningful connection between attachments and tasks but I see how DQL joins would make that much nicer.

ErikEverson committed:
Update index.ts to have types

Co-authored-by: Vincent Ahrend <123380273+pvditto@users.noreply.github.com>

ErikEverson committed:
Update index.ts with better example

Co-authored-by: Vincent Ahrend <123380273+pvditto@users.noreply.github.com>

ErikEverson committed:
Fixed await syntax for disable sync with v3

Co-authored-by: Vincent Ahrend <123380273+pvditto@users.noreply.github.com>

ErikEverson committed:
Added log call when fetch complete

Co-authored-by: Vincent Ahrend <123380273+pvditto@users.noreply.github.com>

ErikEverson committed:
Fixed syntax issues with close

Co-authored-by: Vincent Ahrend <123380273+pvditto@users.noreply.github.com>

ErikEverson committed:
Added error handling

Co-authored-by: Vincent Ahrend <123380273+pvditto@users.noreply.github.com>

ErikEverson committed:
npm install ts-node

ErikEverson committed:
Fixed issues with commits from github

ErikEverson committed:
added all fetch cases rather than default

okdistribute commented:
Yeah it would be more realistic if the attachment was attached to a task

ErikEverson committed:
Unified tasks and attachments and fixed command naming

ErikEverson commented:
@okdistribute  @pvditto  I updated it so that the attachments are on the task now. I also updated the commands to not have the --syntax. 
We could do more elaborate command line stuff but that might just make the demo convoluted as it would be more about parsing terminal commands rather than showing off Ditto.

We could add a library like https://www.npmjs.com/package/command-line-args but that might be a bit much.

pvditto commented:
Yes, let's not add any further dependencies for command-parsing and keep this simple.

Just for the record, @ErikEverson and I had a chat yesterday and decided to change up how the `attachments` command works to provide a clear example that doesn't lead to any race issues in a decentralised setting.

okdistribute commented:
Thanks all! Are we good to merge this at all?

pvditto commented:
@okdistribute I think this was meant to be an alternative to #1 so it needs some conflict resolution now that that is merged. Also, there was something about the `attachments` command but it's been a while and I don't remember the specifics. @ErikEverson I can do a re-review and see if I remember what is was if you have capacity to take this up again or if someone else takes over this PR.

ErikEverson commented:
@okdistribute @pvditto Sorry I have been busy with other stuff. There was something we wanted to do with attachments related to how/when they were fetched.

ErikEverson commented:
I will reopen when I have a chance to update this PR
